File: examples/typescript/basic_workflow.ts
#!/usr/bin/env npx tsx
// Basic workflow demonstration - TypeScript GraphQL Client
// Shows core workflow operations using GraphQL API
// Run with: npx tsx examples/typescript/basic_workflow.ts

interface GraphQLResponse<T = any> {
  data?: T;
  errors?: Array<{ message: string; locations?: any[]; path?: any[] }>;
}

interface WorkflowGQL {
  id: string;
  name: string;
  places: string[];
  transitions: TransitionGQL[];
  initialPlace: string;
  createdAt: string;
  updatedAt: string;
}

interface TransitionGQL {
  id: string;
  name?: string;
  fromPlaces: string[];
  toPlace: string;
  conditions: string[];
  description?: string;
}

interface TokenGQL {
  id: string;
  workflowId: string;
  place: string;
  data: any;
  metadata: any;
  createdAt: string;
  updatedAt: string;
  history: HistoryEventGQL[];
}

interface HistoryEventGQL {
  timestamp: string;
  transition: string;
  fromPlace: string;
  toPlace: string;
  data?: any;
}

class CircuitBreakerClient {
  constructor(private baseUrl: string = 'http://localhost:4000') {}

  async graphql<T = any>(query: string, variables?: any): Promise<GraphQLResponse<T>> {
    const response = await fetch(`${this.baseUrl}/graphql`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ query, variables }),
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    return await response.json() as GraphQLResponse<T>;
  }

  async createWorkflow(input: any) {
    const mutation = `
      mutation CreateWorkflow($input: WorkflowDefinitionInput!) {
        createWorkflow(input: $input) {
          id
          name
          places
          transitions {
            id
            fromPlaces
            toPlace
            conditions
          }
          initialPlace
          createdAt
          updatedAt
        }
      }
    `;

    return this.graphql<{ createWorkflow: WorkflowGQL }>(mutation, { input });
  }

  async createToken(input: any) {
    const mutation = `
      mutation CreateToken($input: TokenCreateInput!) {
        createToken(input: $input) {
          id
          workflowId
          place
          data
          metadata
          createdAt
          updatedAt
          history {
            timestamp
            transition
            fromPlace
            toPlace
            data
          }
        }
      }
    `;

    return this.graphql<{ createToken: TokenGQL }>(mutation, { input });
  }

  async fireTransition(input: any) {
    const mutation = `
      mutation FireTransition($input: TransitionFireInput!) {
        fireTransition(input: $input) {
          id
          workflowId
          place
          data
          metadata
          createdAt
          updatedAt
          history {
            timestamp
            transition
            fromPlace
            toPlace
            data
          }
        }
      }
    `;

    return this.graphql<{ fireTransition: TokenGQL }>(mutation, { input });
  }

  async getToken(id: string) {
    const query = `
      query GetToken($id: String!) {
        token(id: $id) {
          id
          workflowId
          place
          data
          metadata
          createdAt
          updatedAt
          history {
            timestamp
            transition
            fromPlace
            toPlace
            data
          }
        }
      }
    `;

    return this.graphql<{ token: TokenGQL | null }>(query, { id });
  }

  async listWorkflows() {
    const query = `
      query ListWorkflows {
        workflows {
          id
          name
          places
          transitions {
            id
            fromPlaces
            toPlace
            conditions
          }
          initialPlace
          createdAt
        }
      }
    `;

    return this.graphql<{ workflows: WorkflowGQL[] }>(query);
  }
}

function logSuccess(message: string) {
  console.log(`✅ ${message}`);
}

function logInfo(message: string) {
  console.log(`ℹ️  ${message}`);
}

function logError(message: string) {
  console.log(`❌ ${message}`);
}

async function main() {
  console.log('🚀 Circuit Breaker Basic Workflow Demo - TypeScript Client');
  console.log('==========================================================');
  console.log();

  const client = new CircuitBreakerClient();

  try {
    // Create a simple application development workflow
    logInfo('Creating Application Development Workflow...');
    
    const workflowInput = {
      name: 'Application Development Process',
      places: ['planning', 'development', 'testing', 'staging', 'production', 'maintenance'],
      transitions: [
        {
          id: 'start_development',
          fromPlaces: ['planning'],
          toPlace: 'development',
          conditions: []
        },
        {
          id: 'submit_for_testing',
          fromPlaces: ['development'],
          toPlace: 'testing',
          conditions: []
        },
        {
          id: 'back_to_development',
          fromPlaces: ['testing'],
          toPlace: 'development',
          conditions: []
        },
        {
          id: 'promote_to_staging',
          fromPlaces: ['testing'],
          toPlace: 'staging',
          conditions: []
        },
        {
          id: 'deploy_to_production',
          fromPlaces: ['staging'],
          toPlace: 'production',
          conditions: []
        },
        {
          id: 'enter_maintenance',
          fromPlaces: ['production'],
          toPlace: 'maintenance',
          conditions: []
        },
        {
          id: 'back_to_planning',
          fromPlaces: ['maintenance'],
          toPlace: 'planning',
          conditions: []
        }
      ],
      initialPlace: 'planning'
    };

    const workflowResult = await client.createWorkflow(workflowInput);
    
    if (workflowResult.errors) {
      logError(`Failed to create workflow: ${workflowResult.errors.map(e => e.message).join(', ')}`);
      return;
    }

    const workflow = workflowResult.data!.createWorkflow;
    logSuccess(`Created workflow: ${workflow.name} (${workflow.id})`);
    logInfo(`Places: ${workflow.places.join(' → ')}`);
    logInfo(`Transitions: ${workflow.transitions.length} defined`);
    console.log();

    // Create a feature development token
    logInfo('Creating Feature Development Token...');
    
    const tokenInput = {
      workflowId: workflow.id,
      initialPlace: 'planning',
      data: {
        featureName: 'User Authentication',
        assignedDeveloper: 'Alice Smith',
        priority: 'high',
        estimatedHours: 40,
        requirements: [
          'Login/logout functionality',
          'Password reset capability',
          'Session management',
          'Security audit'
        ]
      },
      metadata: {
        createdBy: 'project-manager',
        project: 'main-application',
        sprint: 'sprint-2024-01'
      }
    };

    const tokenResult = await client.createToken(tokenInput);
    
    if (tokenResult.errors) {
      logError(`Failed to create token: ${tokenResult.errors.map(e => e.message).join(', ')}`);
      return;
    }

    const token = tokenResult.data!.createToken;
    logSuccess(`Created token: ${token.id}`);
    logInfo(`Feature: ${token.data.featureName}`);
    logInfo(`Developer: ${token.data.assignedDeveloper}`);
    logInfo(`Current place: ${token.place}`);
    console.log();

    // Simulate development lifecycle
    const transitions = [
      { id: 'start_development', description: 'Start Development Phase' },
      { id: 'submit_for_testing', description: 'Submit for Testing' },
      { id: 'back_to_development', description: 'Back to Development (Bug Found)' },
      { id: 'submit_for_testing', description: 'Resubmit for Testing' },
      { id: 'promote_to_staging', description: 'Promote to Staging' },
      { id: 'deploy_to_production', description: 'Deploy to Production' }
    ];

    let currentToken = token;
    
    for (const transition of transitions) {
      logInfo(`Firing transition: ${transition.description}`);
      
      const transitionInput = {
        tokenId: currentToken.id,
        transitionId: transition.id,
        data: {
          timestamp: new Date().toISOString(),
          performedBy: 'automated-system',
          notes: `Executed ${transition.description}`
        }
      };

      const transitionResult = await client.fireTransition(transitionInput);
      
      if (transitionResult.errors) {
        logError(`Failed to fire transition: ${transitionResult.errors.map(e => e.message).join(', ')}`);
        continue;
      }

      currentToken = transitionResult.data!.fireTransition;
      logSuccess(`Transition completed: ${currentToken.place}`);
      
      // Add a small delay to make the demo more realistic
      await new Promise(resolve => setTimeout(resolve, 100));
    }

    console.log();
    logInfo('Development Lifecycle Complete!');
    logInfo(`Final state: ${currentToken.place}`);
    
    // Show history
    console.log();
    logInfo('Complete Development History:');
    currentToken.history.forEach((event, index) => {
      const timestamp = new Date(event.timestamp).toLocaleTimeString();
      console.log(`  ${index + 1}. ${event.fromPlace} → ${event.toPlace} via ${event.transition} (${timestamp})`);
    });

    console.log();
    logInfo('Workflow demonstrates:');
    console.log('  • Complex state transitions with cycles');
    console.log('  • Rich token data for application features');
    console.log('  • Complete audit trail of state changes');
    console.log('  • GraphQL API integration from TypeScript');
    console.log('  • Production-ready development workflow');

  } catch (error) {
    logError(`Demo failed: ${error}`);
    process.exit(1);
  }
}

if (require.main === module) {
  main().catch(console.error);
}

export { CircuitBreakerClient, type WorkflowGQL, type TokenGQL, type HistoryEventGQL }; 

File: examples/rust/basic_workflow.rs
// Basic workflow example demonstrating GraphQL API usage
// This shows how Rust applications should interact via GraphQL for consistency
// Run with: cargo run --example basic_workflow

use serde_json::json;

// GraphQL endpoint
const GRAPHQL_ENDPOINT: &str = "http://localhost:4000/graphql";

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("🔄 Circuit Breaker - Basic Workflow Demo (GraphQL)");
    println!("==================================================");
    println!("Demonstrating: Rust → GraphQL → NATS Backend");
    println!("(Same API as TypeScript, Python, etc.)");
    println!();

    // Create HTTP client
    let client = reqwest::Client::new();

    // 1. Create a document review workflow via GraphQL
    println!("📋 Creating Document Review Workflow via GraphQL...");
    let workflow_definition = json!({
        "name": "Document Review Process",
        "states": [
            "init", "processing", "review", "approved", "published", "rejected"
        ],
        "activities": [
            {
                "id": "start_processing",
                "fromStates": ["init"],
                "toState": "processing",
                "conditions": []
            },
            {
                "id": "submit_for_review",
                "fromStates": ["processing"],
                "toState": "review",
                "conditions": []
            },
            {
                "id": "approve",
                "fromStates": ["review"],
                "toState": "approved",
                "conditions": []
            },
            {
                "id": "reject",
                "fromStates": ["review"],
                "toState": "rejected",
                "conditions": []
            },
            {
                "id": "publish",
                "fromStates": ["approved"],
                "toState": "published",
                "conditions": []
            }
        ],
        "initialState": "init"
    });

    let create_workflow_query = json!({
        "query": r#"
            mutation CreateWorkflow($input: WorkflowDefinitionInput!) {
                createWorkflow(input: $input) {
                    id
                    name
                    states
                    activities {
                        id
                        fromStates
                        toState
                    }
                    initialState
                }
            }
        "#,
        "variables": {
            "input": workflow_definition
        }
    });

    let workflow_result: serde_json::Value = client
        .post(GRAPHQL_ENDPOINT)
        .json(&create_workflow_query)
        .send()
        .await?
        .json()
        .await?;

    if let Some(errors) = workflow_result.get("errors") {
        println!("❌ Failed to create workflow: {}", errors);
        return Ok(());
    }

    let workflow = &workflow_result["data"]["createWorkflow"];
    let workflow_id = workflow["id"].as_str().unwrap();
    println!(
        "✅ Created workflow: {} ({})",
        workflow["name"], workflow_id
    );
    println!("   States: {:?}", workflow["states"]);
    println!(
        "   Activities: {}",
        workflow["activities"].as_array().unwrap().len()
    );
    println!();

    // 2. Create a document resource via GraphQL
    println!("📄 Creating document resource...");
    let resource_data = json!({
        "title": "Technical Documentation",
        "author": "Engineering Team",
        "document_type": "specification",
        "priority": "high",
        "content": "This is a technical document that needs review.",
        "word_count": 150
    });

    let create_resource_query = json!({
        "query": r#"
            mutation CreateResource($input: ResourceCreateInput!) {
                createResource(input: $input) {
                    id
                    workflowId
                    state
                    data
                    metadata
                    createdAt
                }
            }
        "#,
        "variables": {
            "input": {
                "workflowId": workflow_id,
                "initialState": "init",
                "data": resource_data,
                "metadata": {
                    "created_by": "rust-demo",
                    "department": "engineering"
                }
            }
        }
    });

    let resource_result: serde_json::Value = client
        .post(GRAPHQL_ENDPOINT)
        .json(&create_resource_query)
        .send()
        .await?
        .json()
        .await?;

    if let Some(errors) = resource_result.get("errors") {
        println!("❌ Failed to create resource: {}", errors);
        return Ok(());
    }

    let resource = &resource_result["data"]["createResource"];
    let resource_id = resource["id"].as_str().unwrap();
    println!("✅ Created resource: {}", resource_id);
    println!("   Title: {}", resource["data"]["title"]);
    println!("   Current state: {}", resource["state"]);
    println!();

    // 3. Execute activities via GraphQL
    println!("⚡ Executing document review workflow...");
    let activities = vec![
        ("start_processing", "Start Processing"),
        ("submit_for_review", "Submit for Review"),
        ("approve", "Approve Document"),
        ("publish", "Publish Document"),
    ];

    let mut current_resource_id = resource_id.to_string();

    for (activity_id, description) in activities {
        println!("   ➡️  {} ({})", description, activity_id);

        let execute_activity_query = json!({
            "query": r#"
                mutation ExecuteActivity($input: ActivityExecuteInput!) {
                    executeActivity(input: $input) {
                        id
                        state
                        workflowId
                        data
                        history {
                            timestamp
                            activity
                            fromState
                            toState
                        }
                    }
                }
            "#,
            "variables": {
                "input": {
                    "resourceId": current_resource_id,
                    "activityId": activity_id,
                    "data": {
                        "timestamp": chrono::Utc::now().to_rfc3339(),
                        "performed_by": "rust-demo"
                    }
                }
            }
        });

        match client
            .post(GRAPHQL_ENDPOINT)
            .json(&execute_activity_query)
            .send()
            .await?
            .json::<serde_json::Value>()
            .await?
        {
            activity_result => {
                if let Some(data) = activity_result.get("data") {
                    let updated_resource = &data["executeActivity"];
                    println!("   ✅ New state: {}", updated_resource["state"]);

                    // Simulate processing time
                    tokio::time::sleep(tokio::time::Duration::from_millis(200)).await;
                } else if let Some(errors) = activity_result.get("errors") {
                    println!("   ❌ Activity failed: {}", errors[0]["message"]);
                    break;
                }
            }
        }
    }

    // 4. Query final resource state
    println!();
    println!("📊 Querying final resource state...");
    let get_resource_query = json!({
        "query": r#"
            query GetResource($id: String!) {
                resource(id: $id) {
                    id
                    state
                    data
                    history {
                        timestamp
                        activity
                        fromState
                        toState
                    }
                }
            }
        "#,
        "variables": {
            "id": current_resource_id
        }
    });

    let final_result: serde_json::Value = client
        .post(GRAPHQL_ENDPOINT)
        .json(&get_resource_query)
        .send()
        .await?
        .json()
        .await?;

    if let Some(final_resource) = final_result["data"]["resource"].as_object() {
        println!("✅ Final resource state:");
        println!("   ID: {}", final_resource["id"]);
        println!("   State: {}", final_resource["state"]);
        println!("   Title: {}", final_resource["data"]["title"]);

        if let Some(history) = final_resource["history"].as_array() {
            println!("   History ({} events):", history.len());
            for (i, event) in history.iter().enumerate() {
                println!(
                    "     {}. {} → {} via {}",
                    i + 1,
                    event["fromState"],
                    event["toState"],
                    event["activity"]
                );
            }
        }
    }

    println!();
    println!("🎯 Basic Workflow Demo demonstrates:");
    println!("   • Workflow creation via GraphQL API");
    println!("   • Resource lifecycle management");
    println!("   • Activity execution with state transitions");
    println!("   • Audit trail and history tracking");
    println!("   • Consistent API across all languages");
    println!("   • NATS-backed persistent storage");

    Ok(())
}
